linera-views/src/scylla_db.rs

```diff
@@ -0,0 +1,297 @@
+        for row in rows.into_typed::<(Vec<u8>,)>() {
+            let key = row.unwrap();
+            let short_key = key.0[len..].to_vec();
+            keys.push(short_key);
+        }
+    }
+    Ok(keys)
+}
+
+async fn find_key_values_by_prefix_internal(
+    client: &ScyllaDbClientPair,
+    key_prefix: Vec<u8>,
+) -> Result<Vec<(Vec<u8>, Vec<u8>)>, ScylladbContextError> {
+    let session = &client.0;
+    let namespace = &client.1;
+    // Read the value of a key
+    let len = key_prefix.len();
+    let rows = match get_upper_bound_option(&key_prefix) {
+        None => {
+            let values = (namespace.to_vec(), key_prefix);
+            let query = "SELECT k FROM kv.pairs WHERE namespace = ? AND k >= ?";
+            session.query(query, values).await?
+        }
+        Some(upper_bound) => {
+            let values = (namespace.to_vec(), key_prefix, upper_bound);
+            let query = "SELECT k,v FROM kv.pairs WHERE namespace = ? AND k >= ? AND k < ?";
+            session.query(query, values).await?
+        }
+    };
+    let mut key_values = Vec::new();
+    if let Some(rows) = rows.rows {
+        for row in rows.into_typed::<(Vec<u8>, Vec<u8>)>() {
+            let key = row.unwrap();
+            let short_key = key.0[len..].to_vec();
+            key_values.push((short_key, key.1));
+        }
+    }
+    Ok(key_values)
+}
+
+#[async_trait]
+impl KeyValueStoreClient for ScyllaDbClient {
+    const MAX_CONNECTIONS: usize = 10;
+    type Error = ScylladbContextError;
+    type Keys = Vec<Vec<u8>>;
+    type KeyValues = Vec<(Vec<u8>, Vec<u8>)>;
+
+    async fn read_key_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        let client = self.client.read().await;
+        read_key_internal(&client, key.to_vec()).await
+    }
+
+    async fn read_multi_key_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, Self::Error> {
+        // There is probably a better way in ScyllaDB for downloading several keys than this one.
+        let client = self.client.read().await;
+        let mut values = Vec::new();
+        for key in keys {
+            let value = read_key_internal(&client, key.to_vec()).await?;
+            values.push(value);
+        }
+        Ok(values)
+    }
+
+    async fn find_keys_by_prefix(&self, key_prefix: &[u8]) -> Result<Self::Keys, Self::Error> {
+        let client = self.client.read().await;
+        find_keys_by_prefix_internal(&client, key_prefix.to_vec()).await
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let client = self.client.read().await;
+        find_key_values_by_prefix_internal(&client, key_prefix.to_vec()).await
+    }
+
+    async fn write_batch(&self, batch: Batch, _base_key: &[u8]) -> Result<(), Self::Error> {
+        let client = self.client.read().await;
+        write_batch_internal(&client, batch).await
```

Comment:
What is the point of moving out the code into a top-level function?

Reply:
Ok, I put those functions in the `ScyllaDbClient`.